xbmc/network/upnp/UPnPServer.cpp

```diff
@@ -272,8 +272,9 @@ PLT_MediaObject* CUPnPServer::Build(const std::shared_ptr<CFileItem>& item,
 
     m_logger->debug("Preparing upnp object for item '{}'", (const char*)path);
 
-    if (path == "virtualpath://upnproot") {
+    if (path.StartsWith("virtualpath://upnproot")) {
         path.TrimRight("/");
+        item->m_bIsFolder =  true;
         if (path.StartsWith("virtualpath://")) {
```

Comment:
this condition here doesn't make much sense as it's guarded already by `virtualpath://upnproot` probably some old leftover

Reply:
Looking at the rest of the code, was this functionally backwards previously?  In which case should this if should be testing for requesting the root, in which case this should be the
`if (path == "virtualpath://upnproot") {`

Reply:
Yes it is only testing root (and the whole codeblock below proves it). I think the first condition should be checking for virtualpath:// generically (not sure what other virtualpaths:// exist besides upnproot)

Reply:
I found this list in one of Montellese's posts:

    virtualpath://upnproot/
    musicdb://
    videodb://
    library://video
    sources://video
    special://musicplaylists
    special://profile/playlists
    special://videoplaylists
    addons://sources
    upnp://
    plugin://

In which case virtualpath://upnproot/ would be the root. So perhaps the intent is that the first IF finds the directory and sets it as such. The second IF should be testing IF it is in fact the root and not some subdirectory, and set it as such. So perhaps the intent was:

        if (path.StartsWith("virtualpath://upnproot")) {
            path.TrimRight("/");
            item->m_bIsFolder =  true;
            if (path == "virtualpath://upnproot")) {

In the case of virtualpath://upnproot/sub1/item.txt it would hit the first IF, the path would be found as virtualpath://upnproot/sub1 and get flagged as a folder. The second IF would be false as it is not the root.

In the case of virtualpath://upnproot both IF blocks would be true and it would be set appropriately.

Reply:
Further down there is also something wrong.

    if (object) {
        // remap Root virtualpath://upnproot/ to id "0"
        if (object->m_ObjectID == "virtualpath://upnproot/")
            object->m_ObjectID = "0";
        // remap Parent Root virtualpath://upnproot/ to id "0"
        if (object->m_ParentID == "virtualpath://upnproot/")
            object->m_ParentID = "0";
    }

Reply:
Not really, this last block checks Object_ID and Parent_ID, they refer to different properties. If any of them is `"virtualpath://upnproot/"` kodi needs to remap it to "0". The first request any client does is usually request Object Id "0":

<img width="579" alt="image" src="https://user-images.githubusercontent.com/7375276/221424632-f8a39c5f-96b2-410b-b9c5-36f56356b959.png">

If we would answer this with `<container id="virtualpath://upnproot/"` instead of `<container id="0"` that would be invalid from the client perspective.

Reply:
> virtualpath://

None of the items you point out is a virtualpath:// apart from the first one. We have some old references in the code though:
<img width="826" alt="image" src="https://user-images.githubusercontent.com/7375276/221424987-45fd1226-f821-4349-bd9b-3796109d7ab7.png">

This might be completely broken by now...I doubt we can return any valid data. So I think we're mostly on the same line of though here. First if checks for "virtualpath://" generically (not upnproot) and sets the isFolder property. The nested if checks explicitly for `virtualpath://upnproot`.
But I guess only @Montellese might know for sure, if he still remembers ofc :)

Reply:
> Not really, this last block checks Object_ID and Parent_ID, they refer to different properties. If any of them is `"virtualpath://upnproot/"` kodi needs to remap it to "0". The first request any client does is usually request Object Id "0":
> 
> <img alt="image" width="579" src="https://user-images.githubusercontent.com/7375276/221424632-f8a39c5f-96b2-410b-b9c5-36f56356b959.png">
> 
> If we would answer this with `<container id="virtualpath://upnproot/" parentID=...` instead of `<container id="0" parentID=...` that would be invalid from the client perspective.

Ah, sorry, completely missed noticing it being Object vs Parent.

Reply:
TBH I don't remember. IIRC the `ObjectID`s are used by UPnP clients which provide fixed root elements like "Video", "Music", "Pictures" etc. IIRC we used to map these onto special paths like `virtualpath://upnpvieo` or something like that. But I don't remember if this was removed.

I also checked the UPnP ContentDirectory specification and it only mentions that `ObjectID == 0` is reserved for root.

Reply:
OK I tracked the changes to 370df02d29596b90b57ebba30112353cc363c7f7 and that was one or two years before my time so no wonder I don't remember ;-)

Reply:
nice one, goes in line with my thoughts. I'll fix the if condition later as those virtual paths no longer apply, will take that opportunity to apply the clang-format changes too since the methods are wrongly indented (and result on a huge diff).
Thanks!